File: cpp/ex2.cpp
#include <iostream>
#include "ex2.hpp"

namespace variables{
  double A[jmax][imax];
};


int main(){
  using namespace std;
  using namespace variables;
  setA();
  cout<<"A[0][0]="<<A[0][0]<<endl;
  double B[jmax][imax];
  
  for(int j=0;j<jmax;j++){
  for(int i=0;i<imax;i++){
    B[j][i] =A[j][i];
  }
  }
  cout<<"B[0][0]="<<B[0][0]<<endl;
  changeB(B);
  cout<<"B[0][0]="<<B[0][0]<<endl;
}

void setA(){
  using namespace variables;
  for(int j=0;j<jmax;j++){
  for(int i=0;i<imax;i++){
    A[j][i] =1.0e0;
  }
  }
}

template <size_t imax, size_t jmax>
void changeB(double (&B)[jmax][imax]){
  using namespace variables;
  for(int j=0;j<jmax;j++){
  for(int i=0;i<imax;i++){
    B[j][i] =2.0e0;
  }
  }
}

File: Fortran/Basic77/ex2.f
      program main
      use resolution
      use variables
      implicit none
      real(8),dimension(:,:),allocatable:: B
      imax=2
      jmax=3
      allocate(A(imax,jmax)) 
      call SetA
      write(6,*)"A[1][1]=",A(1,1)
      allocate(B(imax,jmax)) 
      B(:,:) = A (:,:)
      write(6,*)"B[1][1]=",B(1,1)
      call changeB(B)
      write(6,*)"B[1][1]=",B(1,1)
      end program

      subroutine setA()
      use resolution
      use variables
      implicit none
      integer::i,j
      do j=1,jmax
      do i=1,imax
        A(i,j) = 1.0d0
      enddo
      enddo
      return
      end subroutine setA

      subroutine changeB(B)
      use resolution
      implicit none
      real(8),dimension(imax,jmax),intent(inout):: B
      integer::i,j
      do j=1,jmax
      do i=1,imax
        B(i,j) = 2.0d0
      enddo
      enddo
      return
      end subroutine changeB
